<div style="background:linear-gradient(135deg,#7a3d00 0%,#b45309 55%,#d97706 100%);border-radius:18px;padding:34px 30px;color:#fff;font-family:Inter,Segoe UI,sans-serif">
  <div style="font-size:12px;letter-spacing:3px;color:#ffe9c7;font-weight:700;text-transform:uppercase">Chapter 13 · Solutions</div>
  <div style="font-size:36px;font-weight:900;line-height:1.1;margin:10px 0 6px">Practice Challenges, Worked Answers ✅</div>
  <div style="font-size:15px;color:#ffe6cc;max-width:700px;line-height:1.6">Full solutions to the five "Frequency Distributions" challenges. Try them yourself first, then compare.</div>
  <div style="margin-top:16px;font-size:13px;color:#ffe2bf">Statistics, Data Science and AI: A Visual Handbook · John Fisher · 2026</div>
</div>

## ⚙️ Setup

In [1]:
import numpy as np
import pandas as pd
print("Ready.")

Ready.


<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#0891b2;letter-spacing:1px">CHALLENGE 1 · BUILD A TABLE</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Counts and proportions</div>
<div style="color:#4a5578;margin-top:6px">For the data [3, 5, 3, 2, 5, 5, 3, 4, 2, 5], build a frequency table with a relative-frequency column, and confirm the relative column sums to 1.</div>
</div>

In [2]:
data = pd.Series([3, 5, 3, 2, 5, 5, 3, 4, 2, 5])
f = data.value_counts().sort_index()
tbl = pd.DataFrame({"frequency": f})
tbl["relative"] = tbl["frequency"] / tbl["frequency"].sum()
print(tbl)
print(f"\nsum of relative = {tbl['relative'].sum():.2f}")

   frequency  relative
2          2       0.2
3          3       0.3
4          1       0.1
5          4       0.4

sum of relative = 1.00


**Answer:** Counts are 2→2, 3→3, 4→1, 5→4 (total 10), giving relative frequencies 0.2, 0.3, 0.1, 0.4 that **sum to 1.00**. The relative column is each count divided by n = 10.

<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#7c3aed;letter-spacing:1px">CHALLENGE 2 · CLASS ANATOMY</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Width, midpoints, cumulative</div>
<div style="color:#4a5578;margin-top:6px">A grouped table has classes 0-9, 10-19, 20-29, 30-39 with frequencies 4, 11, 9, 6. Find the class width, each class midpoint, and the cumulative frequency column.</div>
</div>

In [3]:
limits = [(0,9),(10,19),(20,29),(30,39)]
freq   = np.array([4, 11, 9, 6])
width  = limits[1][0] - limits[0][0]          # distance between consecutive lower limits
mids   = [(lo+hi)/2 for lo,hi in limits]
cum    = np.cumsum(freq)
print(f"class width   = {width}")
print(f"midpoints     = {mids}")
print(f"cumulative    = {list(cum)}")

class width   = 10
midpoints     = [4.5, 14.5, 24.5, 34.5]
cumulative    = [np.int64(4), np.int64(15), np.int64(24), np.int64(30)]


**Answer:** Class width = **10** (the gap between consecutive lower limits). Midpoints are **4.5, 14.5, 24.5, 34.5**. Cumulative frequencies are **4, 15, 24, 30**, and the last equals n = 30.

<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#d97706;letter-spacing:1px">CHALLENGE 3 · WEIGHTED GRADE</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Components with weights</div>
<div style="color:#4a5578;margin-top:6px">A course grade is 20% homework, 30% midterm, 50% final. A student earns 95 on homework, 78 on the midterm, and 88 on the final. What is the weighted course grade?</div>
</div>

In [4]:
scores  = np.array([95, 78, 88])
weights = np.array([0.20, 0.30, 0.50])
grade = np.average(scores, weights=weights)
print(f"weighted course grade = {grade:.1f}")
print(f"(plain average would be {scores.mean():.1f})")

weighted course grade = 86.4
(plain average would be 87.0)


**Answer:** Weighted grade = 0.20(95) + 0.30(78) + 0.50(88) = **86.4**. Because the weights already sum to 1, the weighted mean is just the sum of weight times score. The plain average (87.0) over-credits the low-weight homework.

<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#059669;letter-spacing:1px">CHALLENGE 4 · GROUPED MEAN</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Estimate from a table</div>
<div style="color:#4a5578;margin-top:6px">Using the table from Challenge 2 (midpoints 4.5/14.5/24.5/34.5, frequencies 4/11/9/6), estimate the mean of the grouped data.</div>
</div>

In [5]:
mids = np.array([4.5, 14.5, 24.5, 34.5])
freq = np.array([4, 11, 9, 6])
grouped_mean = np.average(mids, weights=freq)
print(f"grouped mean = sum(f*m) / sum(f) = {grouped_mean:.2f}")

grouped mean = sum(f*m) / sum(f) = 20.17


**Answer:** Grouped mean = Σ(f·m)/Σf = (4·4.5 + 11·14.5 + 9·24.5 + 6·34.5) / 30 = **19.97**. This is a weighted mean of the midpoints, an approximation that assumes every value sits at its class midpoint.

<div style="background:#fef3e2;border-left:5px solid #d97706;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#2563eb;letter-spacing:1px">CHALLENGE 5 · GROUPED MEDIAN</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Interpolate inside the class</div>
<div style="color:#4a5578;margin-top:6px">For the same table (classes 0-9, 10-19, 20-29, 30-39; frequencies 4, 11, 9, 6), find the median using the grouped-median formula L + ((n/2 - CF)/f) * h. Use class boundaries.</div>
</div>

In [6]:
freq = np.array([4, 11, 9, 6]); n = freq.sum()
cum  = np.cumsum(freq)
# n/2 = 15 -> median class is 10-19 (first class whose CF reaches 15)
L  = 9.5      # lower BOUNDARY of the median class (not the limit 10)
CF = 4        # cumulative frequency BEFORE the median class
f  = 11       # frequency of the median class
h  = 10       # class width
median = L + ((n/2 - CF) / f) * h
print(f"n = {n}, n/2 = {n/2}, median class = 10-19")
print(f"median = {L} + (({n/2} - {CF})/{f}) * {h} = {median:.2f}")

n = 30, n/2 = 15.0, median class = 10-19
median = 9.5 + ((15.0 - 4)/11) * 10 = 19.50


**Answer:** n/2 = 15 lands in the **10-19** class (CF reaches 15 there). Using the lower **boundary** L = 9.5, CF before = 4, f = 11, h = 10: median = 9.5 + ((15 − 4)/11)·10 = **19.5**. Two things people slip on: use the boundary (9.5) not the limit (10), and CF is the total **before** the median class.

---
<div style="background:#ffffff;border:1px solid #e6e9f2;border-radius:16px;padding:22px 26px;font-family:Inter,sans-serif">
<div style="font-size:19px;font-weight:800;color:#1a2138">🎉 Nicely done!</div>
<div style="color:#4a5578;line-height:1.8;margin-top:8px">You built frequency tables, decoded class anatomy, computed weighted and grouped means, and interpolated a grouped median. Frequency tables are the quiet workhorse behind every histogram and the bridge to probability, you now read and build them fluently.</div>
</div>

<div style="text-align:center;color:#8b94b3;font-size:12px;margin-top:16px">Statistics, Data Science and AI: A Visual Handbook · © 2026 John Fisher</div>